packages/astro/src/prefetch/index.ts

```diff
@@ -0,0 +1,260 @@
+/*
+  NOTE: Do not add any dependencies or imports in this file so that it can load quickly in dev.
+*/
+
+// eslint-disable-next-line no-console
+const debug = import.meta.env.DEV ? console.debug : undefined;
+const inBrowser = import.meta.env.SSR === false;
+const prefetchedUrls = new Set<string>();
+const listenedAnchors = new WeakSet<HTMLAnchorElement>();
+
+// User-defined config for prefetch
+// @ts-expect-error injected by vite-plugin-prefetch
+let prefetchAll: boolean = __PREFETCH_PREFETCH_ALL__;
+// @ts-expect-error injected by vite-plugin-prefetch
+let defaultStrategy: string = __PREFETCH_DEFAULT_STRATEGY__;
+
+interface InitOptions {
+	defaultStrategy?: string;
+	prefetchAll?: boolean;
+}
+
+let inited = false;
+/**
+ * Initialize the prefetch script, only works once
+ */
+export function init(opts?: InitOptions) {
+	if (!inBrowser) return;
+
+	prefetchAll ??= opts?.prefetchAll ?? false;
+	defaultStrategy ??= opts?.defaultStrategy ?? 'hover';
+
+	// Init only once
+	if (inited) return;
+	inited = true;
+
+	debug?.(`[astro] Initializing prefetch script`);
+
+	// In the future, perhaps we can enable treeshaking specific unused strategies
+	initTapStrategy();
+	initHoverStrategy();
+	initViewportStrategy();
+}
+
+/**
+ * Prefetch links with higher priority when the user taps on them
+ */
+function initTapStrategy() {
+	for (const event of ['touchstart', 'mousedown']) {
+		document.body.addEventListener(
+			event,
+			(e) => {
+				if (elMatchesStrategy(e.target, 'tap')) {
+					prefetch(e.target.href, { with: 'fetch' });
+				}
+			},
+			{ passive: true, once: true }
```

Comment:
once? So after the first tap we no longer prefetch anything? I am probably misunderstanding....

Reply:
Oh good catch. I thought this was listening on individual links. Thanks I'll remove this.